File: sdk/kotlin/ExampleJWT.kt
/**
 * C2M API Kotlin SDK Example with JWT Authentication
 */

package com.c2m.api.example

import com.c2m.api.apis.AuthApi
import com.c2m.api.apis.DefaultApi
import com.c2m.api.infrastructure.ApiClient
import com.c2m.api.infrastructure.ClientException
import com.c2m.api.infrastructure.ServerException
import com.google.gson.Gson
import kotlinx.coroutines.*
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import java.time.Instant
import java.time.temporal.ChronoUnit

class C2MClient(
    private val clientId: String,
    private val clientSecret: String,
    private val baseUrl: String = "https://api.c2m.com/v2"
) {
    private val httpClient = OkHttpClient()
    private val gson = Gson()
    private var longToken: String? = null
    private var shortToken: String? = null
    private var tokenExpiry: Instant? = null

    data class TokenResponse(
        val access_token: String,
        val token_id: String,
        val expires_at: String
    )

    private suspend fun getLongToken(): String = withContext(Dispatchers.IO) {
        val requestBody = mapOf(
            "grant_type" to "client_credentials",
            "client_id" to clientId,
            "client_secret" to clientSecret,
            "scopes" to listOf("jobs:submit", "jobs:read", "templates:read"),
            "ttl_seconds" to 2592000
        )

        val request = Request.Builder()
            .url("$baseUrl/auth/tokens/long")
            .post(
                gson.toJson(requestBody).toRequestBody("application/json".toMediaType())
            )
            .build()

        httpClient.newCall(request).execute().use { response ->
            if (!response.isSuccessful) {
                throw IOException("Failed to get long token: ${response.code}")
            }

            val responseBody = response.body?.string() ?: throw IOException("Empty response")
            val tokenResponse = gson.fromJson(responseBody, TokenResponse::class.java)
            
            longToken = tokenResponse.access_token
            println("Long-term token obtained: ${tokenResponse.token_id}")
            tokenResponse.access_token
        }
    }

    private suspend fun getShortToken(): String = withContext(Dispatchers.IO) {
        if (longToken == null) {
            getLongToken()
        }

        val requestBody = mapOf(
            "scopes" to listOf("jobs:submit")
        )

        val request = Request.Builder()
            .url("$baseUrl/auth/tokens/short")
            .header("Authorization", "Bearer $longToken")
            .post(
                gson.toJson(requestBody).toRequestBody("application/json".toMediaType())
            )
            .build()

        httpClient.newCall(request).execute().use { response ->
            if (!response.isSuccessful) {
                throw IOException("Failed to get short token: ${response.code}")
            }

            val responseBody = response.body?.string() ?: throw IOException("Empty response")
            val tokenResponse = gson.fromJson(responseBody, TokenResponse::class.java)
            
            shortToken = tokenResponse.access_token
            tokenExpiry = Instant.parse(tokenResponse.expires_at)
            println("Short-term token obtained: ${tokenResponse.token_id}")
            tokenResponse.access_token
        }
    }

    private suspend fun ensureAuthenticated(): String {
        val now = Instant.now()
        if (shortToken == null || tokenExpiry == null || now.isAfter(tokenExpiry?.minus(1, ChronoUnit.MINUTES))) {
            return getShortToken()
        }
        return shortToken!!
    }

    private suspend fun createApiClient(): ApiClient {
        val token = ensureAuthenticated()
        
        return ApiClient(baseUrl).apply {
            addDefaultHeader("Authorization", "Bearer $token")
        }
    }

    suspend fun submitJob(jobParams: Map<String, Any>): Any {
        return try {
            val apiClient = createApiClient()
            val api = DefaultApi(apiClient)
            api.submitSingleDocWithTemplateParams(jobParams)
        } catch (e: ClientException) {
            if (e.statusCode == 401 || e.statusCode == 403) {
                // Token expired, refresh and retry
                shortToken = null
                val apiClient = createApiClient()
                val api = DefaultApi(apiClient)
                api.submitSingleDocWithTemplateParams(jobParams)
            } else {
                throw e
            }
        }
    }
}

// Usage example
fun main() = runBlocking {
    try {
        val client = C2MClient(
            clientId = "your-client-id",
            clientSecret = "your-client-secret"
        )

        val jobParams = mapOf(
            "templateId" to "standard-letter",
            "documentUrl" to "https://example.com/document.pdf",
            "recipients" to listOf(
                mapOf(
                    "name" to "John Doe",
                    "address" to mapOf(
                        "line1" to "123 Main St",
                        "city" to "New York",
                        "state" to "NY",
                        "zip" to "10001"
                    )
                )
            )
        )

        val result = client.submitJob(jobParams)
        println("Job submitted: $result")
        
    } catch (e: Exception) {
        println("Error: ${e.message}")
    }
}

File: sdk/php/example_jwt.php
<?php
/**
 * C2M API PHP SDK Example with JWT Authentication
 */

require_once(__DIR__ . '/vendor/autoload.php');

use C2M\Api\AuthApi;
use C2M\Api\DefaultApi;
use C2M\Configuration;
use C2M\ApiException;
use GuzzleHttp\Client;

class C2MClient {
    private $clientId;
    private $clientSecret;
    private $baseUrl;
    private $authUrl;
    private $longToken;
    private $shortToken;
    private $tokenExpiry;
    private $httpClient;
    
    public function __construct($clientId, $clientSecret, $baseUrl = 'https://api.c2m.com/v2') {
        $this->clientId = $clientId;
        $this->clientSecret = $clientSecret;
        $this->baseUrl = $baseUrl;
        $this->authUrl = $baseUrl;
        $this->httpClient = new Client(['timeout' => 30]);
    }
    
    private function getLongToken() {
        $response = $this->httpClient->post($this->authUrl . '/auth/tokens/long', [
            'json' => [
                'grant_type' => 'client_credentials',
                'client_id' => $this->clientId,
                'client_secret' => $this->clientSecret,
                'scopes' => ['jobs:submit', 'jobs:read', 'templates:read'],
                'ttl_seconds' => 2592000 // 30 days
            ]
        ]);
        
        $data = json_decode($response->getBody(), true);
        $this->longToken = $data['access_token'];
        echo "Long-term token obtained: " . $data['token_id'] . "\n";
        return $this->longToken;
    }
    
    private function getShortToken() {
        if (!$this->longToken) {
            $this->getLongToken();
        }
        
        $response = $this->httpClient->post($this->authUrl . '/auth/tokens/short', [
            'headers' => [
                'Authorization' => 'Bearer ' . $this->longToken
            ],
            'json' => [
                'scopes' => ['jobs:submit']
            ]
        ]);
        
        $data = json_decode($response->getBody(), true);
        $this->shortToken = $data['access_token'];
        $this->tokenExpiry = new DateTime($data['expires_at']);
        echo "Short-term token obtained: " . $data['token_id'] . "\n";
        return $this->shortToken;
    }
    
    private function ensureAuthenticated() {
        $now = new DateTime();
        $buffer = clone $now;
        $buffer->modify('+1 minute');
        
        if (!$this->shortToken || !$this->tokenExpiry || $buffer >= $this->tokenExpiry) {
            $this->getShortToken();
        }
        
        return $this->shortToken;
    }
    
    private function createApiClient() {
        $token = $this->ensureAuthenticated();
        
        $config = Configuration::getDefaultConfiguration();
        $config->setHost($this->baseUrl);
        $config->setAccessToken($token);
        
        return new GuzzleHttp\Client();
    }
    
    public function submitJob($jobParams) {
        $apiInstance = new DefaultApi(
            $this->createApiClient(),
            Configuration::getDefaultConfiguration()->setAccessToken($this->ensureAuthenticated())
        );
        
        try {
            $result = $apiInstance->submitSingleDocWithTemplateParams($jobParams);
            return $result;
        } catch (ApiException $e) {
            if ($e->getCode() == 401 || $e->getCode() == 403) {
                // Token expired, refresh and retry
                $this->shortToken = null;
                $apiInstance = new DefaultApi(
                    $this->createApiClient(),
                    Configuration::getDefaultConfiguration()->setAccessToken($this->ensureAuthenticated())
                );
                return $apiInstance->submitSingleDocWithTemplateParams($jobParams);
            }
            throw $e;
        }
    }
}

// Usage example
try {
    $client = new C2MClient(
        'your-client-id',
        'your-client-secret'
    );
    
    $jobParams = [
        'templateId' => 'standard-letter',
        'documentUrl' => 'https://example.com/document.pdf',
        'recipients' => [[
            'name' => 'John Doe',
            'address' => [
                'line1' => '123 Main St',
                'city' => 'New York',
                'state' => 'NY',
                'zip' => '10001'
            ]
        ]]
    ];
    
    $result = $client->submitJob($jobParams);
    echo "Job submitted: " . json_encode($result) . "\n";
    
} catch (Exception $e) {
    echo 'Error: ' . $e->getMessage() . "\n";
}